rfcs/2022-04-using-native-dialog.md

```diff
@@ -0,0 +1,102 @@
+# RFC: Integrating the native dialog element into existing components
+
+**Date**: April 20, 2022
+**Status**: Ready for review
+
+## Problem
+
+Within the Design System, `<Dialog>` (modal) and `<Drawer>` currently rely on third-party libraries to assist with focus trapping and to manage component elevation. These libraries are:
+
+- [`react-aria-modal`](https://www.npmjs.com/package/react-aria-modal)
+- [`focus-trap-react`](https://www.npmjs.com/package/focus-trap-react)
+
+Relying on third-party libraries to facilitate what can be accomplished natively by the browser presents several pain points:
+
+- The inclusion of libraries increases the bundle size of the Design System, impacting performance for the User
+- The Design System must work within the confines of libraries' APIs, restricting the UX potential of these components
+- Libraries update frequently
+  - Updates may not be compatible with the Design System's current code environment, resulting in using out-of-date libraries
+  - Implementing new library versions are time-consuming and may introduce unexpected behaviors to Design System consuming teams
+- Maintaining these components have additional overhead as Engineers must learn that library's specific API before implementing required changes
+
+## Proposal
+
+The releases of [Firefox (v98)](https://www.mozilla.org/en-US/firefox/98.0/releasenotes/) and [Safari (v15.4)](https://webkit.org/blog/12445/new-webkit-features-in-safari-15-4/) introduce native support for the HTML `<dialog>` element. Chrome and Edge have supported this element for several years, meaning the `<dialog>` element is now [supported across all modern browsers](https://caniuse.com/?search=dialog).
+
+A robust [polyfill](https://github.com/GoogleChrome/dialog-polyfill), maintained by Chrome, is available to ensure support of `<dialog>` for browsers that aren't actively being maintained (IE) or haven't yet been updated (older versions of Firefox and Safari).
+
+Given the pain points listed above and the overlap of functionality between `<Dialog>` and `<Drawer>`, it's suggested the Design System lessen its reliance on third-party libraries by using the native `<dialog>` element for these components instead.
+
+The recommended implementation is to create one internal wrapping `<_Dialog>` component, which houses the `<dialog>` element and provides props that control styling and User interaction (i.e., if background content remains interactive). This component would also manage elevation and focus trapping behavior.
```

Comment:
Would the internal `_Dialog` contain any default styling? Or would that be left to the exported `<Dialog>` & `<Drawer>` to handle?

Reply:
Yeah ☝️ that's good to call out.